Makefile

```diff
@@ -0,0 +1,127 @@
+#!make
+
+.DEFAULT_GOAL := default
+
+default:
+	@echo "Please see README.md for usage of make commands"
+
+##############################################################################
+# Development commands
+##############################################################################
+
+restart: close build run
+
+run:
+	@echo "+\n++ Running client and server...\n+"
+	@docker-compose up -d
+
+close:
+	@echo "+\n++ Stopping client and server...\n+"
+	@docker-compose down
+
+build:
+	@echo "+\n++ Building images...\n+"
+	@docker-compose build
+
+clean:
+	@echo "+\n++ Removing containers, images, volumes etc...\n+"
+	@echo "+\n++ Note: does not clear image cache. \n+"
+	@docker-compose rm -f -v -s
+	@docker volume rm -f openshift-launchpad_postgres-data
+
+client-test:
+	@echo "+\n++ Running client unit tests...\n+"
+	@docker-compose run client npm test
+
+server-test:
+	@echo "+\n++ Running server unit tests...\n+"
+	@docker-compose run server python manage.py test
+
+##############################################################################
+# Deployment / CI-CD commands
+##############################################################################
+
+create-nsp:
+	test -n "$(NAMESPACE)" # Please provide a namespace via NAMESPACE=myproject
+	test -n "$(APP_NAME)" # Please provide an app name via APP_NAME=openshift-launchpad
+	@echo "+\n++ Set network security policies \n+"
+	@oc process -f deployment/nsp.json -p NAMESPACE=$(NAMESPACE) APP_NAME=$(APP_NAME) | oc apply -f -
+
+create-database:
+	test -n "$(NAMESPACE)" # Please provide a namespace via NAMESPACE=myproject
+	test -n "$(APP_NAME)" # Please provide an app name via APP_NAME=openshift-launchpad
+	test -n "$(POSTGRESQL_DATABASE)" # Please provide a database name via POSTGRESQL_DATABASE=sample_db
+	@echo "+\n++ Creating OpenShift database build config and image stream...\n+"
+	@oc process -f deployment/database.bc.json -p NAMESPACE=$(NAMESPACE) APP_NAME=$(APP_NAME) | oc apply -f -
+	@echo "+\n++ Creating OpenShift database deployment config, services, and routes...\n+"
+	@oc process -f deployment/database.dc.json -p NAMESPACE=$(NAMESPACE) APP_NAME=$(APP_NAME) POSTGRESQL_DATABASE=$(POSTGRESQL_DATABASE) | oc apply -f -
```

Comment:
One thing we wanted to avoid in this project was too much obfuscation of the underlying OpenShift CLI commands. 

Obviously, robust functionality is valuable, but this is primarily a tool that can help new lab teams learn OpenShift fundamentals.

Reply:
If that linked file is hard to read, all it's doing is
```sh
# start a build from config
BUILD_OBJECT="$(oc -n "$NAMESPACE" start-build deployment/database.bc.json --output=name)"

# define a function to get the status of the build by returned $BUILD_OBJECT
get_build_status() {
    oc -n "$NAMESPACE" get "$BUILD_OBJECT" -o=go-template='{{.status.phase}}'
}

# wait for the build to be done
while [[ $(get_build_status) == "New" || $(get_build_status) == "Pending" || $(get_build_status) == "Running" ]]; do
    sleep 5
done

# throw an error if the build wasn't successful
if [[ $(get_build_status) != "Complete" ]]; then
    exit 1
fi
```

If you don't trigger & wait for a build, chances are the deployment config will not succeed (esp. on first run) or will not run what you think it will.

Reply:
We are using IS tags to trigger the deployment. While that happens the route will point at an OS "Application is not available" page.

`"triggers": [
          {
            "imageChangeParams": {
              "automatic": true,
              "containerNames": [
                "${APP_NAME}-server"
              ],
              "from": {
                "kind": "ImageStreamTag",
                "name": "${APP_NAME}-server:latest",
                "namespace": "${NAMESPACE}"
              }
            },
            "type": "ImageChange"
          },
          {
            "type": "ConfigChange"
          }
        ],`

Reply:
image change deployments is pretty common so i'm liking this. Interested in how it works with a pipeline backing it